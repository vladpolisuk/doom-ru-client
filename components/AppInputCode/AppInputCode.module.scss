.app_code {
	display: flex;
	align-items: center;
	justify-content: center;
	flex-direction: row;
	gap: 8px;
	margin-bottom: 30px;
}

.app_inputCode {
	display: flex;
	align-items: center;
	justify-content: center;
	text-align: center;
	font-weight: bold;
	width: 50px;
	height: 50px;
	font-size: 20px;
	padding: 0;
	border: 1px solid var(--color-6);
	background-color: var(--background_actives);

	&::-webkit-outer-spin-button,
	&::-webkit-inner-spin-button {
		-webkit-appearance: none;
		margin: 0;
	}

	&[type='number'] {
		-moz-appearance: textfield;
	}

	&:focus::placeholder {
		color: transparent;
	}
}
