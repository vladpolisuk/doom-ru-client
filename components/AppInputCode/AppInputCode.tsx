import {
	BaseSyntheticEvent,
	ChangeEvent,
	ClipboardEvent,
	FC,
	InputHTMLAttributes,
	KeyboardEvent,
	memo,
	useEffect,
	useState
} from 'react';
import { BaseAppComponent } from '../../types/components';
import AppInput from '../AppInput/AppInput';
import s from './AppInputCode.module.scss';

type IAppInputCode = BaseAppComponent<HTMLInputElement> &
	InputHTMLAttributes<HTMLInputElement> & {
		codeLength: number;
		codePlaceholder?: string;
		submit: (code: number) => void;
	};

/**
 * The common input code component in the application
 * @type `FC<AppInputCode>`
 * @memo `true`
 * @return `html:input[]`
 */
const AppInputCode: FC<IAppInputCode> = memo(props => {
	const {
		title,
		children,
		type,
		codeLength,
		placeholder,
		className = '',
		onlyARIA = false,
		name = 'codeNumber',
		codePlaceholder = '𐤏',
		submit,
		...extra
	} = props;
	const [code, setCode] = useState('');

	useEffect(() => {
		if (code.length === codeLength) submit(+code);
	}, [code, codeLength, submit]);

	const onChange = (event: ChangeEvent<HTMLInputElement>) => {
		if (!event.target.value) return;
		const id = +event.target.id;
		const value = +event.target.value;
		const fields = document.getElementsByName(name) as NodeListOf<HTMLInputElement>;
		if (id !== codeLength - 1) fields.item(id + 1).focus();
		const div = Math.floor(value / 10);
		if (div < 1) return setCode(code + value);
		fields.item(id).value = String(div);
	};

	const handleKeyDown = (event: BaseSyntheticEvent<HTMLInputElement> & KeyboardEvent<HTMLInputElement>) => {
		const id = +event.target.id;
		if (event.target.value || event.key !== 'Backspace') return;
		const fields = document.getElementsByName(name) as NodeListOf<HTMLInputElement>;
		if (id === 0) return;
		fields.item(id - 1).focus();
		setCode(code.slice(0, code.length - 1));
	};

	const onPaste = (event: BaseSyntheticEvent<HTMLInputElement> & ClipboardEvent<HTMLInputElement>) => {
		event.preventDefault();
		const id = +event.target.id;
		const fields = document.getElementsByName(name) as NodeListOf<HTMLInputElement>;
		const paste = event.clipboardData?.getData('text');
		if (!paste) return;
		const pasteArray = paste.split('');
		if (pasteArray.length > codeLength - id) return;
		pasteArray.forEach((value, index) => {
			fields.item(index).value = value;
		});
		setCode(paste);
	};

	return (
		<div className={s.app_code}>
			{[...Array(codeLength).keys()].map(value => (
				<AppInput
					name={name}
					type='number'
					maxLength={1}
					id={`${value}`}
					key={value}
					data-testid='app-inputCode'
					onChange={onChange}
					onPaste={onPaste}
					onKeyDown={handleKeyDown}
					placeholder={codePlaceholder}
					className={s.app_inputCode}
					{...extra}
				/>
			))}
		</div>
	);
});

AppInputCode.displayName = 'AppInputCode';
export default AppInputCode;
