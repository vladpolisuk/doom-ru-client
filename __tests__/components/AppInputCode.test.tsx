import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import AppInputCode from '../../components/AppInputCode/AppInputCode';

describe('AppInputCode', () => {
	it('should render', () => {
		const wrapper = render(<AppInputCode codeLength={6} />);
		const input = screen.getAllByTestId('app-inputCode')[0];
		expect(input).toBeInTheDocument();
		expect(wrapper).toMatchSnapshot();
	});

	it('should not have value', () => {
		render(<AppInputCode codeLength={6} />);
		const inputs = screen.getAllByTestId('app-inputCode');
		inputs.forEach(input => expect(input).toHaveValue(null));
	});

	it('should have value', () => {
		render(<AppInputCode codeLength={6} />);
		const inputs = screen.getAllByTestId('app-inputCode');
		inputs.forEach(input => {
			fireEvent.input(input, { target: { value: 1 } });
		});
		inputs.forEach(input => expect(input).toHaveValue(1));
	});

	it('should have value with paste', () => {
		render(<AppInputCode codeLength={6} />);
		const inputs = screen.getAllByTestId('app-inputCode');
		fireEvent.paste(inputs[0], { clipboardData: { getData: () => '123456' } });
		inputs.forEach((input, index) => expect(input).toHaveValue(index + 1));
	});
});
