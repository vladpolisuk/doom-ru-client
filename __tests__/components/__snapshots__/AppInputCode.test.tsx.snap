// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInputCode should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app_code"
      >
        <input
          class="app_input app_inputCode"
          data-testid="app-inputCode"
          id="0"
          maxlength="1"
          name="codeNumber"
          placeholder="𐤏"
          type="number"
        />
        <input
          class="app_input app_inputCode"
          data-testid="app-inputCode"
          id="1"
          maxlength="1"
          name="codeNumber"
          placeholder="𐤏"
          type="number"
        />
        <input
          class="app_input app_inputCode"
          data-testid="app-inputCode"
          id="2"
          maxlength="1"
          name="codeNumber"
          placeholder="𐤏"
          type="number"
        />
        <input
          class="app_input app_inputCode"
          data-testid="app-inputCode"
          id="3"
          maxlength="1"
          name="codeNumber"
          placeholder="𐤏"
          type="number"
        />
        <input
          class="app_input app_inputCode"
          data-testid="app-inputCode"
          id="4"
          maxlength="1"
          name="codeNumber"
          placeholder="𐤏"
          type="number"
        />
        <input
          class="app_input app_inputCode"
          data-testid="app-inputCode"
          id="5"
          maxlength="1"
          name="codeNumber"
          placeholder="𐤏"
          type="number"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app_code"
    >
      <input
        class="app_input app_inputCode"
        data-testid="app-inputCode"
        id="0"
        maxlength="1"
        name="codeNumber"
        placeholder="𐤏"
        type="number"
      />
      <input
        class="app_input app_inputCode"
        data-testid="app-inputCode"
        id="1"
        maxlength="1"
        name="codeNumber"
        placeholder="𐤏"
        type="number"
      />
      <input
        class="app_input app_inputCode"
        data-testid="app-inputCode"
        id="2"
        maxlength="1"
        name="codeNumber"
        placeholder="𐤏"
        type="number"
      />
      <input
        class="app_input app_inputCode"
        data-testid="app-inputCode"
        id="3"
        maxlength="1"
        name="codeNumber"
        placeholder="𐤏"
        type="number"
      />
      <input
        class="app_input app_inputCode"
        data-testid="app-inputCode"
        id="4"
        maxlength="1"
        name="codeNumber"
        placeholder="𐤏"
        type="number"
      />
      <input
        class="app_input app_inputCode"
        data-testid="app-inputCode"
        id="5"
        maxlength="1"
        name="codeNumber"
        placeholder="𐤏"
        type="number"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
